[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/chunxianlin/CSC448/blob/main/CSC448_HW3.ipynb)

# Part 1.Loading the dataset
## Using pandas load the dataset (load remotely, not locally)
## Output the first 15 rows of the data
## Display a summary of the table information (number of datapoints, etc.)

In [96]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
import numpy as np

In [105]:
#read the csv file
dataset = pd.read_csv("https://raw.githubusercontent.com/profmcnich/example_notebook/main/science_data_large.csv")
#print the first 15 rows of data 
dataset.head(15)

Temperature °C  Mols KCL     Size nm^3
0              469       647  6.244743e+05
1              403       694  5.779610e+05
2              302       975  6.196847e+05
3              779       916  1.460449e+06
4              901        18  4.325726e+04
5              545       637  7.124634e+05
6              660       519  7.006960e+05
7              143       869  2.718260e+05
8               89       461  8.919803e+04
9              294       776  4.770210e+05
10             991       117  2.441771e+05
11             307       781  5.006455e+05
12             206        70  3.145200e+04
13             437       599  5.390215e+05
14             566        75  9.185271e+04

In [106]:
#print the information about the dataset
dataset.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 3 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Temperature °C  1000 non-null   int64  
 1   Mols KCL        1000 non-null   int64  
 2   Size nm^3       1000 non-null   float64
dtypes: float64(1), int64(2)
memory usage: 23.6 KB


# Part 2. Splitting the dataset
## Take the pandas dataset and split it into our features (X) and label (y)

## Use sklearn to split the features and labels into a training/test set. (90% train, 10% test)

In [99]:
#Split the dataset. X = temperature and Mol, y = size
X = dataset[["Temperature °C","Mols KCL"]]
y = dataset["Size nm^3"]
#Split the dataset into train dataset and test dataset
X_train, X_test,y_train, y_test = train_test_split(X,y, test_size=0.1,random_state=10)
X_train.shape,y_train.shape
X_test.shape,y_test.shape

# Part 3. Perform a Linear Regression
## Use sklearn to train a model on the training set

## Create a sample datapoint and predict the output of that sample with the trained model

## Report on the score for that model, in your own words (markdown, not code) explain what the score means

## Extract the coefficents and intercept from the model and write an equation for your h(x) using LaTeX

In [111]:
#Linear regression model for training
reg = LinearRegression()

reg.fit(X_train,y_train)
reg.score(X_test,y_test)

0.8838673863856773

In [ ]:
#the score function will return the coefficient of determination of the prediction. The best possible score is 1.0. In this case the score is 
#0.88 which means the model is accurate for the most of the predictions.

In [110]:
reg.coef_

array([ 876.56037994, 1032.21149885])

In [113]:
reg.intercept_

-420227.3447910782

$$h(x) = 1032.21149885 = 876.56037994m - 420227.3447910782$$
$$m = 480.58247432$$
$$h(x) = 480.58247432x - 420227.3447910782$$

# Part 4. Use Cross Validation
## Use the cross_val_score function to repeat your experiment across many shuffles of the data

## Report on their finding and their significance

In [119]:
from sklearn.model_selection import cross_val_score
print(cross_val_score(reg,X,y,cv=12))

[0.80420866 0.88011209 0.8205926  0.88487535 0.86545715 0.87435564
 0.85205157 0.86396832 0.87573787 0.818712   0.80478287 0.89268586]


In [ ]:
#The lowest value is 0.8042 and the highest is 0.8926. The average value approximately 0.85 which means 
#the initial score is better than the average score. Also the model can be improved

# Part 5. Using Polynomial Regression
## Using the PolynomialFeatures library perform another regression on an augmented dataset of degree 2

## Report on the metrics and output the resultant equation as you did in Part 3.

In [131]:
from sklearn.preprocessing import PolynomialFeatures
poly = PolynomialFeatures(2)
pf = poly.fit_transform(X)

In [132]:
from sklearn import linear_model
poly_reg = linear_model.Lasso(alpha=0.1)
poly_reg.fit(pf,y)

Lasso(alpha=0.1)

In [133]:
poly_reg.predict(pf)

array([6.24472863e+05, 5.77959866e+05, 6.19685132e+05, 1.46044979e+06,
       4.32581645e+04, 7.12461969e+05, 7.00694620e+05, 2.71826851e+05,
       8.91988043e+04, 4.77020514e+05, 2.44178340e+05, 5.00644900e+05,
       3.14532404e+04, 5.39020025e+05, 9.18521941e+04, 3.95303776e+04,
       5.38420799e+05, 1.14869260e+04, 1.48585683e+05, 4.16308659e+05,
       1.31597217e+05, 4.82432059e+05, 1.16136527e+06, 1.36056034e+04,
       4.24488807e+05, 1.97796772e+04, 8.03034676e+05, 3.21294567e+05,
       6.95232444e+05, 2.23961980e+05, 1.10432874e+06, 1.92627576e+06,
       5.21372522e+05, 7.91713976e+05, 4.53953294e+05, 5.11927279e+04,
       5.94763919e+04, 4.40629451e+05, 4.60782309e+05, 7.88617753e+04,
       8.03209308e+05, 2.28364337e+05, 2.41598517e+05, 1.04577978e+06,
       6.59940953e+04, 6.18551360e+04, 3.97634922e+05, 4.93008203e+05,
       1.18456776e+05, 2.42666313e+05, 1.26719051e+06, 1.28495739e+05,
       6.10293664e+05, 5.76107612e+04, 9.13728391e+05, 1.41796390e+06,
      

In [134]:
poly_reg.score(pf,y)

0.9999999999928951

In [135]:
#The score of the this model is 0.999 which is very close to 1. It means the model is very accurate for predicting.

In [136]:
poly_reg.coef_

array([ 0.00000000e+00,  1.19858321e+01, -8.30756262e-03,  1.18383848e-05,
        2.00000307e+00,  2.85784254e-02])

In [137]:
poly_reg.intercept_

4.159558213141281

$$\hat{y}(w,x) = 4.159558213141281 + 0x_1+11.98x_2-0.0083x_3+0.00001x_4+2x_5+0.02858x_6$$